src/lib.rs

```diff
@@ -0,0 +1,414 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A failpoints implementation for rust.
+
+extern crate rand;
+
+use std::str::FromStr;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::{Condvar, Mutex, RwLock, TryLockError};
+use std::thread;
+use std::time::{Duration, Instant};
+
+use rand::Closed01;
+
+/// Supported tasks.
+#[derive(Clone, Debug, PartialEq)]
+enum Task {
+    /// Do nothing.
+    Off,
+    /// Return the value.
+    Return(Option<String>),
+    /// Sleep for some milliseconds.
+    Sleep(u64),
+    /// Panic with the message.
+    Panic(Option<String>),
+    /// Print the message.
+    Print(Option<String>),
+    /// Sleep until other action is set.
+    Pause,
+    /// Yield the CPU.
+    Yield,
+    /// Busy waiting for some milliseconds.
+    Delay(u64),
+}
+
+#[derive(Debug)]
+struct Action {
+    task: Task,
+    freq: f32,
+    count: Option<AtomicUsize>,
+}
+
+impl PartialEq for Action {
+    fn eq(&self, hs: &Action) -> bool {
+        if self.task != hs.task || self.freq != hs.freq {
+            return false;
+        }
+        if let Some(ref lhs) = self.count {
+            if let Some(ref rhs) = hs.count {
+                return lhs.load(Ordering::Relaxed) == rhs.load(Ordering::Relaxed);
+            }
+        } else if hs.count.is_none() {
+            return true;
+        }
+        false
+    }
+}
+
+impl Action {
+    fn new(task: Task, freq: f32, max_cnt: Option<usize>) -> Action {
+        Action {
+            task: task,
+            freq: freq,
+            count: max_cnt.map(AtomicUsize::new),
+        }
+    }
+
+    fn get_task(&self) -> Option<Task> {
```

Comment:
How about `maybe_get_task`? Sin it may not return a task even if there is one.

Reply:
I think `get` is ok. For example, in stdlib, there are many `get` methods return `Option`, but no method named `maybe_get`.